addons/account/account_payment.py

```diff
@@ -0,0 +1,378 @@
+            counterpart_aml = aml_obj.create(counterpart_aml_dict)
+            move.post()
+
+            # Reconcile the invoice, if present
+            if res.invoice_id:
+                if res.payment_difference != 0.0 and res.payment_difference_handling == 'reconcile':
+                    res.invoice_id.register_payment(counterpart_aml, res.writeoff_account, res.journal_id)
+                else:
+                    res.invoice_id.register_payment(counterpart_aml)
+
+            # In case of a transfer, the first journal entry created debited the source liquidity account and credited
+            # the transfer account. Now we debit the transfer account and credit the destination liquidity account.
+            if res.payment_type == 'transfer':
+                dst_move = res.env['account.move'].create(res._get_move_vals(res.destination_journal_id))
+                transfer_debit_aml_dict = res._get_shared_move_line_vals(credit, debit, -amount_currency, dst_move.id)
+                transfer_debit_aml_dict.update({
+                    'name': 'TODO',
+                    'payment_id': res.id,
+                    'account_id': res.env.user.company_id.transfer_account.id,
+                    'journal_id': res.destination_journal_id.id })
+                transfer_debit_aml = aml_obj.create(transfer_debit_aml_dict)
+                dst_liquidity_aml_dict = res._get_shared_move_line_vals(debit, credit, amount_currency, dst_move.id)
+                dst_liquidity_aml_dict.update({
+                    'name': 'TODO',
+                    'account_id': res.destination_journal_id.default_credit_account_id.id,
+                    'journal_id': res.destination_journal_id.id })
+                aml_obj.create(dst_liquidity_aml_dict)
+                dst_move.post()
+
+                (counterpart_aml + transfer_debit_aml).reconcile()
+
+            res.state = 'confirmed'
+            if res.payment_method.code == 'manual':
+                res.payment_state = 'done'
+
+    @api.multi
+    def button_create_and_post(self):
+        # When you click a button, the record is created before the button's method is called.
+        # In action_account_invoice_payment, there is no 'Create' button, so we simulate it with this hack.
+        for rec in self:
+            try:
+                with self._cr.savepoint():
+                    rec.post()
+            except Exception:
+                # If an error occurs in post(), the payment must be deleted.
+                # We use an explicit rollback to revert ORM operations done in post()
+                # then invalidate affected caches, delete the record and propagate the exception
+                self.invalidate_cache()
+                self.env['account.move'].invalidate_cache()
+                self.env['account.move.line'].invalidate_cache()
+
+                rec.unlink()
+                self._cr.commit()
+                raise
+
+    @api.multi
+    def unlink(self):
```

Comment:
replace method name by button_cancel(),  unlink is allowed only when state=='canceled'